File: 883-car-fleet/car-fleet.java
class Solution {
    class Pair{
        int position;
        int speed;
        Pair(int position, int speed){
            this.position = position;
            this.speed = speed;
        }
    }
    double timeTaken(int position, int speed, int target) {
        return (double) (target - position) / speed;
    }
    public int carFleet(int target, int[] position, int[] speed) {
        Pair[] pairs = new Pair[speed.length];
        for(int i = 0; i < position.length; i++){
            pairs[i] = new Pair(position[i], speed[i]);
        }
        Arrays.sort(pairs, new Comparator<Pair>() {
            @Override
            public int compare(Pair p1, Pair p2) {
                return p2.position - p1.position; 
            }
        });
        Stack<Pair> stack = new Stack<>();
        int i = 0;
        int res = 0;
        while(i < pairs.length){
            double time = timeTaken(pairs[i].position, pairs[i].speed, target);
            if (!stack.isEmpty() && time <= timeTaken(stack.peek().position, stack.peek().speed, target)) {
                i++;
                continue;
            }
            res++;
            stack.push(pairs[i]);
            i++;
        }
        return res;
    }
}

File: 883-car-fleet/car-fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        tuplearr = []
        stack = []
        def helper(target, position, speed):
            return (target-position)/speed
        for i in range(0, len(position)):
            tuplearr.append((-position[i], speed[i]))
        tuplearr.sort()
        for element in tuplearr:
            if stack and stack[-1] >= helper(target, -element[0], element[1]):
                continue
            else:
                stack.append(helper(target, -element[0], element[1]))
        return len(stack)




        